Ia rotation fixes

ianyimi opened the pull request:
rotation changes:
- after clicking rotate

- you default rotating along y

- the active axis has a visual line running through it

- you can switch with keyboard shortcuts x/y/z (experiment with tab to rotate through) (have visual indicator for shortcut)

- increase sensitivity

- hold cmd/ctrl for fine alignment (1º) (normally 10º)

- rotate along global coordinates

alex-shortt commented:
some issues from a quick qa pass:

- using ctrl to get fine movement seems to get enabled and stay enabled even after letting go
- initial click on rotation causes it to snap to a specific rotation, and then works from there
- the axis of rotation is not the actual axis of rotation, you need to rotate along the global axes rather than the global axes. this means you will probably set x y and z all at the same time when rotating along one global axis
- should have caught this earlier, but `<` and `>` should snap to nearest 90º, not add or subtract it
- sensitivity is just a bit too high imo. although yeah that alone is much better
- you're using the camera's global z rotation for x/z rotation. i can tell because looking along the x axis, i can rotate fine, but looking along the z axis nothing happens.

ianyimi commented:
> * using ctrl to get fine movement seems to get enabled and stay enabled even after letting go
changed to a toggle, press to enable, press again to disable
> * initial click on rotation causes it to snap to a specific rotation, and then works from there
this has been a thing. i'm not sure how to fix it yet
> * the axis of rotation is not the actual axis of rotation, you need to rotate along the global axes rather than the global axes. this means you will probably set x y and z all at the same time when rotating along one global axis
currently works this way
> * should have caught this earlier, but `<` and `>` should snap to nearest 90º, not add or subtract it
fixed
> * sensitivity is just a bit too high imo. although yeah that alone is much better
lowered
> * you're using the camera's global z rotation for x/z rotation. i can tell because looking along the x axis, i can rotate fine, but looking along the z axis nothing happens.
fixed

alex-shortt commented on src/realities/BuilderReality/tools/BuilderWand/components/Rotate/components/DisplayAxis.tsx:
```diff
@@ -0,0 +1,66 @@
+type AxisSelectorProps = {
+  active: boolean;
+  activeAxis: "x" | "y" | "z";
+  setActiveAxis: React.Dispatch<React.SetStateAction<"x" | "y" | "z">>;
```
replace `React.Dispatch<React.SetStateAction<"x" | "y" | "z">>` with `(a: "x" | "y' | "z") => void`

alex-shortt commented on src/realities/BuilderReality/tools/BuilderWand/components/Rotate/components/DisplayAxis.tsx:
```diff
@@ -0,0 +1,66 @@
+    const axisMesh = new Mesh(axisGeo, axisMat)
+    axisMesh.name = "rotation-line-mesh"
+    axisMesh.position.set(idea.position.x, idea.position.y, idea.position.z);
+    // axisMesh.rotation.copy(idea.rotation);
```
remove comment

alex-shortt commented on src/realities/BuilderReality/tools/BuilderWand/components/Rotate/components/DisplayAxis.tsx:
```diff
@@ -0,0 +1,66 @@
+  const { active, activeAxis, setActiveAxis } = props;
+  const { scene } = useThree();
+  const { idea } = useFocus();
+
```
should be good for now, but add in the code

`//TODO: https://github.com/pmndrs/react-three-fiber/blob/master/markdown/recipes.md#re-parenting`

alex-shortt commented on src/realities/BuilderReality/tools/BuilderWand/components/Rotate/index.tsx:
```diff
@@ -80,7 +97,9 @@ export function Rotate(props: GroupProps) {
-      objectRot.copy(idea.rotation.toVector3());
+      const worldQuaternion = new THREE.Quaternion();
+      idea.getWorldQuaternion(worldQuaternion);
+      objectRot.set(worldQuaternion.x, worldQuaternion.y, worldQuaternion.z);
```
`objectRot.setFromQuaternion(worldQuaternion)`

alex-shortt reviewed (changes requested):
- ctrl doesn't seem to work for me
- quaternion needs some fixing
- if set exactly to a 90º, then < and > should go to the next/prev 90º rotation
